Device Naming Screens

ErikSin opened the pull request:
Checks if user has named their device, and conditionally renders the device naming screens if that has not happened.

This pr is stacked on #106 

![image](https://github.com/digidem/CoMapeo-mobile/assets/67773827/090f1dbc-087e-4aec-b72f-c0491a67a7b9)

<img width="290" alt="image" src="https://github.com/digidem/CoMapeo-mobile/assets/67773827/d3de78cd-7526-4f54-bdc7-2ec75d87630e">

<img width="290" alt="image" src="https://github.com/digidem/CoMapeo-mobile/assets/67773827/5eb677d3-f867-445d-8e43-61e3bff4bf00">

![image](https://github.com/digidem/CoMapeo-mobile/assets/67773827/2da6c2b2-3869-4d5a-b4bb-e77f7b3b4f29)

achou11 commented:
Do you mind updating this to point to the main branch? sorry i asked you to update it so that it was stacked recently 😅

ErikSin commented:
> Do you mind updating this to point to the main branch? sorry i asked you to update it so that it was stacked recently 😅

Its odd it didn't do that automatically when we merged it...I swear that is usually what happens.

achou11 commented:
> Its odd it didn't do that automatically when we merged it...I swear that is usually what happens.

yeah usually it does but i guess since you manually changed the branch before, it decided not to auto update this time around.

looks like you may need to rebase your branch onto main (since it used to be stacked on a PR that has since been merged)...

achou11 reviewed (changes requested):
are the screenshots accurate? seems like the copy for the naming screen is different than what's implemented

mostly comments around translations. can re-review when rebase issues are resolved :)

ErikSin commented:
> are the screenshots accurate? seems like the copy for the naming screen is different than what's implemented

The screens shots are from before I added the appropriate copy from the copy deck

achou11 commented:
> The screens shots are from before I added the appropriate copy from the copy deck

ah ok. i find it more helpful to at least include screenshots that match the implementation - or both if you're okay with the extra work.

ErikSin commented on src/frontend/screens/IntroToCoMapeo.tsx:
```diff
@@ -0,0 +1,95 @@
+import {DeviceNamingSceens} from '../Navigation/ScreenGroups/DeviceNamingScreens';
+
+const m = defineMessages({
+  isNow: {
+    id: 'screens.IntroToCoMapeo.isNow',
+    defaultMessage: 'is now',
+    description: "The full sentence is mapeo 'is now' comapeo. ",
+  },
```
No, unfortuantly "Mapeo" and "CoMapeo" are svgs. CoMapeo is the "logo". I suppose we could turn "Mapeo" into text, but we would still have the same problems

achou11 replied:
ah right - thanks for the clarification

ErikSin commented on src/frontend/screens/DeviceNaming.tsx:
```diff
@@ -0,0 +1,148 @@
+  description: {
+    id: 'screens.DeviceNaming.description',
+    defaultMessage:
+      'To create or join a new project you will have to name your device. In a future version of CoMapeo, you will be able to manage the projects you created by changing devices roles, deleting devices and much more.',
```
This is copied directly from the copy deck. I think we should bring that up there instead

ErikSin commented:
> > The screens shots are from before I added the appropriate copy from the copy deck
> 
> ah ok. i find it more helpful to at least include screenshots that match the implementation - or both if you're okay with the extra work.

I just updated them

achou11 commented on src/frontend/screens/DeviceNaming.tsx:
```diff
@@ -0,0 +1,148 @@
+  );
+};
+
+function useErrorTimeout() {
+  const [errorTimeout, setErrorTimeout] = React.useState(false);
+  const timer = React.useRef<NodeJS.Timeout | undefined>();
+
+  React.useEffect(() => {
+    if (errorTimeout && !timer.current) {
+      timer.current = setTimeout(() => {
+        setErrorTimeout(false);
+        timer.current = undefined;
+      }, 1500);
+    }
+
+    return () => {
+      if (timer.current) {
+        clearTimeout(timer.current);
+        timer.current = undefined;
+      }
+    };
+  }, [errorTimeout]);
+
+  return [
+    errorTimeout,
+    () => setErrorTimeout(prevVal => (!prevVal ? true : prevVal)),
+  ] as const;
+}
```
Think this could be implemented in a way that:

1. Avoids effects that rely on state changes and instead colocates them with explicit event handlers/callbacks
2. Generalizes the value that's ephemeral
3. Allows configurable time length

Ended up with this, which is similar to `useState` in terms of usage with a couple of extra things:

1. An option to configure duration before resetting to the initial value
2. A third element in the return value which can be used to reset to the initial state before the duration is reached (e.g. clearing an error state immediately based on some subsequent user interaction)

```ts
// initialValue and resetAfter usually won't change too much, as they are meant to be like configuration options, similar to how the built-in React hooks work
function useEphemeralState<T>(initialValue: T, resetAfter: number) {
  const [value, setValue] = React.useState<T>(initialValue);

  const timer = React.useRef<NodeJS.Timeout | undefined>();

  // There's probably something I'm missing here in terms of resetting some state or cleaning up when initialValue or resetAfter change, but not sure how consequential that is
  React.useEffect(() => {
    return () => {
      if (timer.current) {
        clearTimeout(timer.current);
        timer.current = undefined;
      }
    };
  }, []);

  function set(value: Parameters<typeof setValue>[0]) {
    setValue(value);
    
    // co-locate the restart/creation of the timer with the event handler/callback, based on React idioms
    if (timer.current) {
      clearTimeout(timer.current);
    }

    timer.current = setTimeout(() => {
      setValue(initialValue);
    }, resetAfter);
    

  }

  function reset() {
    setValue(initialValue);

    if (timer.current) {
      clearTimeout(timer.current);
      timer.current = undefined;
    }
  }

  return [value, set, reset] as const;
}
```

In this component, usage would be like this (although):

```ts
// `error` is inferred to be boolean here, so the explicit generic isn't technically needed in this case
const [error, setError] = useEphemeralState<boolean>(false, 1500);

function handleNamePress() {
  if (invalidName) {
    setError(true)
  }
  
  ...
}
```

Not adamant about using this change exactly as is, but I think it would be worth considering to avoid dealing with more complicated useEffects

ErikSin replied:
I agree, the useEffect is not ideal. But I think instead of implementing custom validation hooks we should look into libraries (eg. react hook forms). I am going to leave it as is, but earmark it as something we should address

achou11 reviewed (approved):
Things are coming together! Added a suggestion regarding one of the hooks you've implemented, but wouldn't consider it blocking.

Non-blocking - and not sure if it's my device (Pixel 2) - but noticing some UI inconsistencies related to text size and weight:

- "Collaborate..." text should have increased font weight
- "Get Started" button font size looks a bit small

<img width="300" alt="image" src="https://github.com/digidem/CoMapeo-mobile/assets/18542095/399c87f2-b00d-40c4-a078-f1a8faf62bf2">

- description text in grey box should have greater font weight
- "Add Name" button text looks a bit small

<img width="300" alt="image" src="https://github.com/digidem/CoMapeo-mobile/assets/18542095/e408855e-3295-43fe-8df6-7e98eb063d56">

Probably fine to address as a follow-up given some of these are tied to shared components. I haven't been too strict about styling but it's ideally something to properly address at some point. Can create an issue if that's helpful
